x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java

```diff
@@ -64,6 +66,8 @@
     private final Map<String, Map<Step.StepKey, Step>> stepMap;
     private final NamedXContentRegistry xContentRegistry;
 
+    private final Map<Index, Tuple<IndexMetadata, Step>> cachedSteps = new ConcurrentHashMap<>();
```

Comment:
Lets add a line or two of Javadoc explaining what this is for please. This is far from obvious for anybody not involved in this work.

Reply:
How does 5f48e35b29ff8d9945bc5c14735409e0b84c549e sparkle with you?

Reply:
>  IndexMetadata feels really heavyweight for this

We only use it for instance equality and clear it out of the cache if the index gets removed => memory wise it's free, instance equality comparison is cheap/free and this seems like the safest way of checking whether or not index metadata has changed.

We didn't just create a compound key here because we need/want the ability to quickly cleanup the cache if an index is removed without having to iterate a cache. Since we only care about the by index state this seemed fastest/easiest.

Reply:
This has the side effect however, or any inconsequential change to the index metadata for the index (a dynamic field introduced, for example) invalidating the cache, whereas it doesn't actually need to be invalidated in that case?

Reply:
I think that's ok. This is just supposed to help us with not running the expensive step computation on unchanged indices (which will always be the majority in almost any CS update). No need to super optimize this beyond that I'd say. Updating index metadata results in expensive operations anyway that even go as far as resulting in disk IO, the step calculation is probably trivial relatively speaking in all cases.

Reply:
Alright, that's fine then, my preference is still to make the keyed part as small as possible, but this is okay for now I think.